### PR TITLE
Add files via upload

### PR DESCRIPTION
**Pull request by AhsanAkbar252**

### CODE REVIEW

**Comment by suleman-uzair** [comment]
Add some description in PR, what's the PR adding to the application

File: Shape.rb [hunk 1]
```diff
@@ -0,0 +1,86 @@
+class Shape
```
Comment (suleman-uzair) [inline Shape.rb:1]: make separate files for separate classes

File: Shape.rb [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  def getArea
+  end
```
Comment (suleman-uzair) [inline Shape.rb:3]: Read about abstract classes in ruby

File: Shape.rb [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+  def getPerimeter
```
Comment (suleman-uzair) [inline Shape.rb:5]: As Shape might be an abstract class, we can add exception in its method
Method not implemented

File: Shape.rb [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  def hasEdges
+    true
+  end
+
+  def sides
+  end
+end
+
+class Rectangle < Shape
+  def initialize(h, w)
```
Comment (suleman-uzair) [inline Shape.rb:17]: read about attribute accessors, attribute readers , attribute writers in ruby classes

Comment (suleman-uzair) [inline Shape.rb:17]: use descriptive variable names

File: Shape.rb [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+  def getArea
+    @height * @width
+  end
+
+  def getPerimeter
+    2*@height + 2*@width
+  end
+
+  def sides
+    return 4
```
Comment (suleman-uzair) [inline Shape.rb:31]: ruby methods return the last used value, no need of return here

File: Shape.rb [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+  def initialize(h, w)
+    @height = h
+    @width = w 
+  end
+
+  def getArea
+    @height * @width
+  end
+
+  def getPerimeter
```
Comment (suleman-uzair) [inline Shape.rb:26]: camel_case method names

File: Shape.rb [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    2*@height + 2*@width
+  end
+
+  def sides
+    return 4
+  end
+end
+
+class Triangle < Shape
+  def initialize(s1, s2, s3)
```
Comment (suleman-uzair) [inline Shape.rb:36]: same as above

File: Shape.rb [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    @side2 = s2
+    @side3 = s3 
+  end
+
+  def getPerimeter
+    @side1 + @side2 + @side3
+  end
+
+  def getArea
+    @s = getPerimeter/2
```
Comment (suleman-uzair) [inline Shape.rb:47]: same as above, descriptive variable names
semi_perimeter

File: Shape.rb [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+  def initialize(s1, s2, s3)
+    @side1 = s1
+    @side2 = s2
+    @side3 = s3 
+  end
+
+  def getPerimeter
+    @side1 + @side2 + @side3
+  end
+
```
Comment (suleman-uzair) [inline Shape.rb:45]: as it is a difficult formula to understand, add it as a code comment

File: Shape.rb [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+  def initialize(r)
+    @radius = r
+  end
+
+  def getPerimeter
+    2 * 3.14 * @radius
+  end
+
+  def getArea
+    3.14 * @radius**2
```
Comment (suleman-uzair) [inline Shape.rb:69]: 2 suggestions, we are using 3.14 at multiple places, we can make it a constant
or use maths library pi value

File: Shape.rb [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+  end
+end
+
+
+
+class Circle < Shape
+  def initialize(r)
+    @radius = r
+  end
+
```
Comment (suleman-uzair) [inline Shape.rb:63]: code comment formula

File: Shape.rb [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+  def getPerimeter
+    2 * 3.14 * @radius
+  end
+
+  def getArea
+    3.14 * @radius**2
+  end
+
+  def hasEdges
```
Comment (suleman-uzair) [inline Shape.rb:72]: boolean methods  have question mark in them
has_edges?

File: Shape.rb [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+    false
+  end
+
+  def sides
+    return 0
+  end
+end
+
+rec = Rectangle.new(4,3)
```
Comment (suleman-uzair) [inline Shape.rb:81]: make a separate file for controller/main class

File: Shape.rb (hunk 4 above)
Comment (AhsanAkbar252) [reply Shape.rb:17]: done

File: Shape.rb [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+    @height = h
+    @width = w 
+  end
+
+  def getArea
+    @height * @width
+  end
+
+  def getPerimeter
+    2*@height + 2*@width
```
[the remark this answers is not in the document]
Comment (AhsanAkbar252) [reply Shape.rb:27]: done

**Pull request merged**